server/services/omnichannel-voip/service.ts

```diff
@@ -1,10 +1,84 @@
-import { ILivechatVoipService } from '../../sdk/types/IOmnichannelVoipService';
+import { Db } from 'mongodb';
+import _ from 'underscore';
+
+import { IOmnichannelVoipService } from '../../sdk/types/IOmnichannelVoipService';
 import { ServiceClass } from '../../sdk/types/ServiceClass';
+import { IVoipConnectorResult } from '../../../definition/IVoipConnectorResult';
+import { IVoipExtensionBase } from '../../../definition/IVoipExtension';
+import { Logger } from '../../lib/logger/Logger';
+import { Voip } from '../../sdk';
+import { IAgentExtensionMap, IOmnichannelVoipServiceResult } from '../../../definition/ILivechatVoipServiceResult';
+/**
+ * Note (Amol) :
+ * Why cant we get the direct access to the model here and have to work with
+ * raw model? This is still a server code. So ideally it should have access to the model
+ * like the REST apis have.
+ *
+ * Check this during the code review and after code review is done, delete this comment.
+ */
+import { UsersRaw } from '../../../app/models/server/raw/Users';
+import { IUser } from '../../../definition/IUser';
 
-export class OmnichannelVoipService extends ServiceClass implements ILivechatVoipService {
+export class OmnichannelVoipService extends ServiceClass implements IOmnichannelVoipService {
 	protected name = 'omnichannel-voip';
 
+	private logger: Logger;
+
+	private users: UsersRaw;
+
+	constructor(db: Db) {
+		super();
+		this.users = new UsersRaw(db.collection('users'));
+		this.logger = new Logger('OmnichannelVoipService');
+	}
+
+	private async getAllocatedExtesionAllocationData(projection: Partial<{ [P in keyof IUser]: number }>): Promise<Array<IUser>> {
+		const roles: string[] = ['livechat-agent', 'livechat-manager', 'admin'];
+		const options = {
+			sort: {
+				username: 1,
+			},
+			projection,
+		};
+
+		const query = {
+			extension: { $exists: true },
+		};
+		return this.users.findUsersInRolesWithQuery(roles, query, options).toArray();
+	}
+
 	getConfiguration(): any {
 		return {};
 	}
+
+	async getAvailableExtensions(): Promise<IOmnichannelVoipServiceResult> {
```

Comment:
How often will this be called/modified? Do you think it's worth to add a small cache to the extension associations to avoid having to go to the db each time? We do have some npm packages to manage timed cache

Reply:
Extension list is coming from Asterisk so I dont think we can cache that. We may think of local allocations to be cached. But I will prefer taking  a conservative approach till we need optimisation.